### DISCUSSION TITLE
Using netplan on WSL2

### DISCUSSION THREAD

**Discussion by horseinthesky** [microsoft/WSL#5354] [General]
**Is your feature request related to a problem? Please describe.**
I've tried this on both Ubuntu 18.04 and 20.04 WSL2.

My `/etc/network/interfaces` on Ubuntu 18.04 says:
```
cat /etc/network/interfaces
# ifupdown has been replaced by netplan(5) on this system.  See
# /etc/netplan for current configuration.
# To re-enable ifupdown on this system, you can run:
#    sudo apt install ifupdown
```
on 20.04 this file is absent.

Netplan is installed on both distros:
18.04
```
apt-cache policy netplan.io
netplan.io:
  Installed: 0.98-0ubuntu1~18.04.1
  Candidate: 0.98-0ubuntu1~18.04.1
  Version table:
 *** 0.98-0ubuntu1~18.04.1 500
        500 http://archive.ubuntu.com/ubuntu bionic-updates/main amd64 Packages
        100 /var/lib/dpkg/status
     0.40.1~18.04.4 500
        500 http://security.ubuntu.com/ubuntu bionic-security/main amd64 Packages
     0.36.1 500
        500 http://archive.ubuntu.com/ubuntu bionic/main amd64 Packages
```
20.04
```
apt-cache policy netplan.io
netplan.io:
  Installed: 0.99-0ubuntu1
  Candidate: 0.99-0ubuntu3~20.04.1
  Version table:
     0.99-0ubuntu3~20.04.1 500
        500 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 Packages
 *** 0.99-0ubuntu1 500
        500 http://archive.ubuntu.com/ubuntu focal/main amd64 Packages
        100 /var/lib/dpkg/status
```

But I cannot use it.
I've created netplan config file:
```
cat /etc/netplan/01-netcfg.yaml
network:
  version: 2
  renderer: networkd
  ethernets:
    eth0:
      dhcp4: true
```

but when I run `sudo netplan apply` it shows me parsing error:
```
sudo netplan apply
/etc/netplan/01-netcfg.yaml:1:9: Error in network definition: expected mapping (check indentation)
network:
        ^
```

**Describe the solution you'd like**
Be able to control networking stack via `netplan` including IPv6.

**Describe alternatives you've considered**
I don't understand how WSL2 distros get IPv4 addresses from HyperV.

**Comment by SerEs35** [comment]
Try `sudo netplan --debug apply` and share the output please.

**Comment by horseinthesky** [comment]
Here it is
```
sudo netplan --debug apply
** (generate:81): DEBUG: 18:22:20.698: Processing input file /etc/netplan/01-netcfg.yaml..
** (generate:81): DEBUG: 18:22:20.731: starting new processing pass
(generate:81): GLib-GIO-DEBUG: 18:22:20.738: _g_io_module_get_default: Found default implementation local (GLocalVfs) for ?gio-vfs?
/etc/netplan/01-netcfg.yaml:1:9: Error in network definition: expected mapping (check indentation)
network:
```

**Comment by WSLUser** [comment]
IPv6 is not supported. There's another issue filed for that already. Localhost works for both WSL2 and Windows as many people wanted that experience from WSL1. Other IPv4 addresses should still work.

**Comment by horseinthesky** [comment]
It was probably related to some spacing issues. I manually deleted all whitespaces and put them again and this is what I get now:
```
sudo netplan --debug apply
** (generate:598): DEBUG: 20:30:37.486: Processing input file /etc/netplan/01-netcfg.yaml..
** (generate:598): DEBUG: 20:30:37.486: starting new processing pass
** (generate:598): DEBUG: 20:30:37.486: eth0: setting default backend to 1
** (generate:598): DEBUG: 20:30:37.486: Configuration is valid
** (generate:598): DEBUG: 20:30:37.487: Generating output files..
** (generate:598): DEBUG: 20:30:37.487: NetworkManager: definition eth0 is not for us (backend 1)
DEBUG:netplan generated networkd configuration changed, restarting networkd
System has not been booted with systemd as init system (PID 1). Can't operate.
Traceback (most recent call last):
  File "/usr/sbin/netplan", line 23, in <module>
    netplan.main()
  File "/usr/share/netplan/netplan/cli/core.py", line 50, in main
    self.run_command()
  File "/usr/share/netplan/netplan/cli/utils.py", line 130, in run_command
    self.func()
  File "/usr/share/netplan/netplan/cli/commands/apply.py", line 44, in run
    self.run_command()
  File "/usr/share/netplan/netplan/cli/utils.py", line 130, in run_command
    self.func()
  File "/usr/share/netplan/netplan/cli/commands/apply.py", line 102, in command_apply
    utils.systemctl_networkd('stop', sync=sync, extra_services=['netplan-wpa@*.service'])
  File "/usr/share/netplan/netplan/cli/utils.py", line 82, in systemctl_networkd
    subprocess.check_call(command)
  File "/usr/lib/python3.6/subprocess.py", line 311, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['systemctl', 'stop', '--no-block', 'systemd-networkd.service', 'netplan-wpa@*.service']' returned non-zero exit status 1.
```

So now it is now working due to `systemd` is not working on WSL?!

**Comment by horseinthesky** [comment]
Another interesting note: I've enabled `systemd` support by running this 
https://github.com/DamionGans/ubuntu-wsl2-systemd-script

After I was able to commit `sudo netplan apply` successfully...well, kind of. It broke WSL2 distro with `systemd` completely AND network stack for all other distros (because all of them somehow and for some reason share ALL network settings). The was neither IPv4 address on `eth0` nor `default` route.

Anyway, after reboot network stack healed itself.

So, results:
1) I can't use `netplan` (without `systemd`).
2) I can't use different IPv4 addresses for various distros (why?).
3) I have no IPv6 connectivity (again, why?).
4) I can't bridge WSL2 distro network interface to Windows system network interface to get IPv4/IPv6 from same subnets (why?).

**Comment by therealkenc** [comment]
> 1.2.3.4.

We can let this float for a bit in case there is some community input to share, but this one just went:

> 2. If I list multiple bugs/concerns in this one issue, the core team may close my issue without further explanation or engagement.

I haven't dove netplan except to look at the [landing page](https://netplan.io/) (because no CLI repro in the OP to cut and paste), but on a broad glance it seems to require as a dependency `network-manager` or `systemd-networkd`, which (to put it lightly) would require a lot of hypothetical CLI steps before one ever made it to "netplan".

**Comment by horseinthesky** [comment]
`netplan` itself should work after `systemd` finally arrive at WSL.

But I still don't have clear understanding regarding what is going with network stack.
2) Why all distros share the same network stack? Is it going to change? Is there a thread for this one?
3) IPv6 discussed here https://github.com/microsoft/WSL/issues/4518 and this one is mandatory for 2020.
4) Bridging is being discussed here https://github.com/microsoft/WSL/issues/4150, but as I understand there are no dates when (or even if) it might be implemented.

**Reply by WSLUser** [reply]
> why all distros share the same network stack? 

Because they all exist in the same microVM. They are using namespace containment to be logically separated from each other. So when you run a container, you are technically running a container within a container, which can also be in a container (if you're using the systemd hack that deals with daemonize).

**Comment by igorMIA** [comment]
Still waiting for netplan support
